src/emulate.c

```diff
@@ -368,17 +370,21 @@ static bool branch_taken = false;
 static uint32_t last_pc = 0;
 
 /* Interpreter-based execution path */
-#define RVOP(inst, code)                                    \
-    static bool do_##inst(riscv_t *rv, const rv_insn_t *ir) \
-    {                                                       \
-        rv->csr_cycle++;                                    \
-        code;                                               \
-    nextop:                                                 \
-        rv->PC += ir->insn_len;                             \
-        if (unlikely(RVOP_NO_NEXT(ir)))                     \
-            return true;                                    \
-        const rv_insn_t *next = ir->next;                   \
-        MUST_TAIL return next->impl(rv, next);              \
+#define RVOP(inst, compressed, code)                                        \
```

Comment:
Maybe we can hardcode PC += 4 and place PC -= 2 in all C-extension instruction.

Reply:
> Maybe we can hardcode PC += 4 and place PC -= 2 in all C-extension instruction.

No, the offset for the next PC should be determined within each instruction. Whether it's 4 or 2 bytes, this value should be predictable and consistent.

Reply:
Because we cannot get the information about the instruction is compressed or not, I only figure two idea, the first is adding this information in parameter list, and the second is handling PC += 2 or 4 per function.

Reply:
> Because we cannot get the information about the instruction is compressed or not, I only figure two ideas, the first is adding this information in parameter list, and the second is handling PC += 2 or 4 per function.

How about extending `src/decode.h` by addressing compressed instructions?

Reply:
However, the RVOP(...) marco is unrelated to `src/decode.h`.

Reply:
> However, the RVOP(...) marco is unrelated to `src/decode.h`.

We can utilize the X-macro trick to enumerate the length of each instruction. For example, `RV_INSN_LEN_addi = 4`. Then, the macro RVOP can directly employ concatenation in the C preprocessor to manage PC increment.